### PR TITLE
Use limine as the bootloader for both UEFI and BIOS.

### PR DESCRIPTION
**Pull request by mobinmob**
This PR implements a solution to https://github.com/void-linux/void-mklive/issues/313 .
It uses limine, a single, well maintained bootloader with a single and simple configuration to replace isolinux and grub. It successfully boots in my tests - suggestions for improvements are welcome!

~As limine is not available, I made a single-package repo for it, signed with my key. Anyone can test this PR by running mklive.sh wiith:~

~`# ./mklive.sh -r https://ftp.halifax.rwthaachen.de/osdn/storage/g/a/av/avyssos/mklive-pkgs`~

~Or build complete isos with build-x86-images:~

~`# ./build-x86-images.sh -r https://ftp.halifax.rwth-aachen.de/osdn/storage/g/a/av/avyssos/mklive-pkgs -b base`~

### CODE REVIEW

**Comment by mobinmob** [comment]
- Rebase on master.

Limine is now available in the repo, so no need for the unofficial one :)

File: mklive.sh.in [hunk 1]
```diff
@@ -277,18 +221,20 @@ generate_squashfs() {
...
-        -omit-version-number -relaxed-filenames -allow-lowercase \
-        -volid "VOID_LIVE" \
-        -eltorito-boot boot/isolinux/isolinux.bin \
-        -eltorito-catalog boot/isolinux/boot.cat \
-        -no-emul-boot -boot-load-size 4 -boot-info-table \
-        -eltorito-alt-boot -e boot/grub/efiboot.img -isohybrid-gpt-basdat -no-emul-boot \
-        -isohybrid-mbr "$SYSLINUX_DATADIR"/isohdpfx.bin \
-        -output "$OUTPUT_FILE" "$IMAGEDIR" || die "Failed to generate ISO image"
-}
+"$VOIDHOSTDIR"/usr/bin/xorriso -as mkisofs \
```
Comment (classabbyamp) [inline mklive.sh.in:224]: indent missing, and an extra space at the start of the following lines?

File: mklive.sh.in [hunk 2]
```diff
@@ -442,3 +387,5 @@ generate_iso_image
 
 hsize=$(du -sh "$OUTPUT_FILE"|awk '{print $1}')
 info_msg "Created $(readlink -f "$OUTPUT_FILE") ($hsize) successfully."
+
+
```
Comment (classabbyamp) [inline mklive.sh.in:391]: extra newlines

File: limine/limine.cfg [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+:@@BOOT_TITLE@@ @@KERNVER@@ @@ARCH@@ (RAM)
+	PROTOCOL=linux
+    KERNEL_PATH=boot:///boot/vmlinuz
+    CMDLINE= root=live:CDLABEL=VOID_LIVE init=/sbin/init ro rd.luks=0 rd.md=0 rd.dm=0 loglevel=4 vconsole.unicode=1 vconsole.keymap=@@KEYMAP@@ locale.LANG=@@LOCALE@@  @@BOOT_CMDLINE@@ rd.live.ram
+    MODULE_PATH=boot:///boot/initrd
+    COMMENT=Boot a Void linux system loaded in RAM
+    
+:Boot next storage device found
+ 	PROTOCOL=chainload-next
+ 	COMMENT=Try to boot from the next bootable storage drive found in the system
```
Comment (classabbyamp) [inline limine/limine.cfg:24]: add a newline at the end of this file

File: limine/limine.cfg [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    MODULE_PATH=boot:///boot/initrd
+    COMMENT=Boot a Void linux live system
+    
+
+:@@BOOT_TITLE@@ @@KERNVER@@ @@ARCH@@ (RAM)
+	PROTOCOL=linux
+    KERNEL_PATH=boot:///boot/vmlinuz
+    CMDLINE= root=live:CDLABEL=VOID_LIVE init=/sbin/init ro rd.luks=0 rd.md=0 rd.dm=0 loglevel=4 vconsole.unicode=1 vconsole.keymap=@@KEYMAP@@ locale.LANG=@@LOCALE@@  @@BOOT_CMDLINE@@ rd.live.ram
+    MODULE_PATH=boot:///boot/initrd
+    COMMENT=Boot a Void linux system loaded in RAM
```
Comment (classabbyamp) [inline limine/limine.cfg:20]:
```suggestion
    COMMENT=Boot a @@BOOT_TITLE@@ system loaded in RAM
```
perhaps?
(and same for other entry)

File: limine/limine.cfg (hunk 3 above)
Comment (classabbyamp) [inline limine/limine.cfg:24]:
```suggestion
 	COMMENT=Try to boot from the next bootable storage device found in the system
```

**Review by classabbyamp — COMMENTED** [review]
how was splash.bmp generated? is it simple enough that it could be done at runtime? if not, should it replace splash.png?

---

not commenting on functionality just yet, would need to test this locally first

**Comment by mobinmob** [comment]
> how was splash.bmp generated? is it simple enough that it could be done at runtime? if not, should it replace splash.png?
> 
It was converted from the png using gimp :p

> not commenting on functionality just yet, would need to test this locally first

**Comment by mobinmob** [comment]
@classabbyamp Thanks for the review, all done :)
The reason I tried my hand at this is that I find limine a better choice that both isolinux and grub for a live cd. Simpler to configure and install for both uefi and bios, single conf file...
extlinux/isolinux is not really developed anymore AFAICT and grub is an unbeliably featurefull and complex beast - a fantastic choice for those that need the features, but an overkill for a live cd.

**Comment by 0x5c** [comment]
I built an `x86_64` live image from this PR. The generated ISO contains a rather odd set of partitions for a live image
```
$ lsblk -f                                             
NAME        FSTYPE    FSVER      LABEL     UUID                                 FSAVAIL FSUSE% MOUNTPOINTS
loop0       iso9660   Joliet Ext VOID_LIVE 2023-02-24-04-38-11-00                              
├─loop0p1   iso9660   Joliet Ext VOID_LIVE 2023-02-24-04-38-11-00                              
├─loop0p2   vfat      FAT12      VOID_LIVE 44E7-1D24                                           
└─loop0p3   iso9660   Joliet Ext VOID_LIVE 2023-02-24-04-38-11-00
```
compared to the last official live image
```
$ lsblk -f                                             
NAME        FSTYPE    FSVER      LABEL     UUID                                 FSAVAIL FSUSE% MOUNTPOINTS
loop0       iso9660   Joliet Ext VOID_LIVE 2022-10-01-17-45-36-00                              
├─loop0p1   iso9660   Joliet Ext VOID_LIVE 2022-10-01-17-45-36-00                              
└─loop0p2   vfat      FAT12      grub_uefi C35E-68B8
```

When checked with cfdisk, the latest image appears to be using a MBR partition table while the Limine one appears to be using GPT. IIRC this would make it non-bootable on BIOS systems. It's also odd how two of the partitions show as `iso9660`.

I tested the live image on an `x86_64` machine using EFI. Limine starts properly, menu/etc works, it loads linux and the initrd just fine, but dracut fails to find the root and switch to it.

```
mount: /run/initramfs/live: unknown filesystem type ''.
       dmesg(1) may have more information after failed mount system call.
[    6.526684] dracut: FATAL: Failed to mount block device of live image
[    6.526686] dracut: Refusing to continue

[... about 5 minutes later ...]

dracut Warning: Could not boot.

dracut Warning: dracut: FATAL: Failed to mount block device of live image
dracut Warning: dracut: Refusing to continue
dracut Warning: "/dev/root" does not exist

[dracut drops to a rescue shell]
```

Astonishingly, booting this live image from grub loopback* works flawlessly, despite there being 3 partitions with the same label it looks for.
*loading the kernel+initrd directly from a bootloader outside the ISO and letting dracut find the iso+squashfs+ext3fs

**Comment by mobinmob** [comment]
Testing with a fresh base iso:

- Boots on virtualbox (both bios and uefi)
- Boots on my desktop pc with ventoy (bios and uefi) 
(I expected the above, since this is the way I am testing this from the start...)
- Fails to boot on my desktop pc with the image written with dd (on both uefi and bios), same error as in https://github.com/void-linux/void-mklive/pull/314#issuecomment-1442886495
- Fails to boot on my laptop (only uefi and dd on this, with a finicky uefi implementation).

`lsblk -f` output from vbox with uefi: https://termbin.com/dx75t
`lsblk -f` output from vbox with bios: https://termbin.com/6rn91

It is entirelly possible I have some error in xorriso when creating the image.

@mintsuki : Any ideas where to look for the cause?

@0x5c Can you check if your image works on vbox?

File: limine/limine.cfg [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+:@@BOOT_TITLE@@ @@KERNVER@@ @@ARCH@@ (RAM)
+	PROTOCOL=linux
+    KERNEL_PATH=boot:///boot/vmlinuz
+    CMDLINE= root=live:CDLABEL=VOID_LIVE init=/sbin/init ro rd.luks=0 rd.md=0 rd.dm=0 loglevel=4 vconsole.unicode=1 vconsole.keymap=@@KEYMAP@@ locale.LANG=@@LOCALE@@  @@BOOT_CMDLINE@@ rd.live.ram
+    MODULE_PATH=boot:///boot/initrd
+    COMMENT=Boot a @@BOOT_TITLE@@ loaded in RAM
+    
+:Boot next storage device found
+ 	PROTOCOL=chainload-next
```
Comment (mintsuki) [inline limine/limine.cfg:23]: The protocol name is `chainload_next`

File: limine/limine.cfg [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+:@@BOOT_TITLE@@ @@KERNVER@@ @@ARCH@@
+    PROTOCOL=linux
+    KERNEL_PATH=boot:///boot/vmlinuz
+    CMDLINE= root=live:CDLABEL=VOID_LIVE init=/sbin/init ro rd.luks=0 rd.md=0 rd.dm=0 loglevel=4 vconsole.unicode=1 vconsole.keymap=@@KEYMAP@@ locale.LANG=@@LOCALE@@  @@BOOT_CMDLINE@@
+    MODULE_PATH=boot:///boot/initrd
+    COMMENT=Boot a @@BOOT_TITLE@@ live system
+    
+
+:@@BOOT_TITLE@@ @@KERNVER@@ @@ARCH@@ (RAM)
+	PROTOCOL=linux
```
Comment (mintsuki) [inline limine/limine.cfg:16]: This is a tab, not spaces.

File: limine/limine.cfg [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    KERNEL_PATH=boot:///boot/vmlinuz
+    CMDLINE= root=live:CDLABEL=VOID_LIVE init=/sbin/init ro rd.luks=0 rd.md=0 rd.dm=0 loglevel=4 vconsole.unicode=1 vconsole.keymap=@@KEYMAP@@ locale.LANG=@@LOCALE@@  @@BOOT_CMDLINE@@ rd.live.ram
```
Comment (mintsuki) [inline limine/limine.cfg:18]: Leading space in command line string.

File: limine/limine.cfg [hunk 8]
```diff
@@ -0,0 +1,24 @@
+INTERFACE_BRANDING=Void Linux
+TIMEOUT=8
+TERM_WALLPAPER=boot:///boot/limine/@@SPLASHIMAGE@@
+TERM_WALLPAPER_STYLE=centered
+TERM_BACKDROP=#000000
+
+:@@BOOT_TITLE@@ @@KERNVER@@ @@ARCH@@
+    PROTOCOL=linux
+    KERNEL_PATH=boot:///boot/vmlinuz
+    CMDLINE= root=live:CDLABEL=VOID_LIVE init=/sbin/init ro rd.luks=0 rd.md=0 rd.dm=0 loglevel=4 vconsole.unicode=1 vconsole.keymap=@@KEYMAP@@ locale.LANG=@@LOCALE@@  @@BOOT_CMDLINE@@
```
Comment (mintsuki) [inline limine/limine.cfg:10]: Leading space in command line string.

**Comment by mintsuki** [comment]
@0x5c GPT images boot on most BIOS systems, the few BIOS systems that I encountered that refuse to boot a ISOHYBRID (which this image is) are actually UEFI capable systems that see that a GPT is present and refuse to boot the image using CSM (which is a non issue since these systems are UEFI capable).

**Comment by mintsuki** [comment]
@mobinmob I'm not sure about that dracut issue, but it doesn't sound Limine-related to me. Perhaps the stray leading spaces on the kernel command lines may have something to do?

**Comment by mobinmob** [comment]
> @mobinmob I'm not sure about that dracut issue, but it doesn't sound Limine-related to me. Perhaps the stray leading spaces on the kernel command lines may have something to do?

Thank you for the review!
Chainload now works on bios.
The other issue is something with dracut, I agree, I am just trying to determine where to start searching.
Booting from usb disk while chainloading the iso by ventoy but now when it is written by dd is weird (exactly same usb drivers, same pc(s).

**Comment by mobinmob** [comment]
@0x5c Many thanks again from a fellow contributor for taking the time to test this PR and give feedback.

> My guess is that either vbox/ventoy/etc are using more fault-tolerant iso9660 parsing, or that they succeed to boot the image because there's no physical volume for the image to not play nice with. Either way, the live images have to be valid once written to physical media.

That is a reasonable conclusion. I tried to find something that will trigger this odd behaviour in dracut with no results - but I am not really familiar with the codebase.

On a slightly unrelated note, the reference of EasyOS by @mintsuki in their limine PR made me revisit the articles about the refusal by Barry Kauler to produce live isos for it, choosing[ to create and release disk image files instead](https://bkhome.org/news/202112/why-iso-was-retired.html). Anyway, that is a seperate issue :) 

Ι will try to find a solution for the issue and close the PR if I fail.

**Comment by 0x5c** [comment]
I'm fairly busy this week, but I'll be exploring possible solutions too

**Comment by mintsuki** [comment]
@0x5c Limine does not *require* any special xorriso incantation. You can generate bootable images (or disks) using Limine a whole number of different ways.

The reference xorriso command (as provided in Limine's own README) produces an ISOHYBRID using GPT + a protective MBR (as well as the actual ISO filesystem). This command is actually very similar to (almost taken right from) the one used by GRUB2 in their `grub-mkrescue` utility.

Given that I have hardly ever heard reports of an ISO generated in such matter not booting on any given system, I am inclined to believe that the modifications to the reference xorriso command made by @mobinmob are (at least partly) to blame, and I'd suggest sticking as closely as possible to the reference xorriso "incantation" to avoid this sort of issues.

Additionally, I would disregard the warnings raised by the Linux kernel w.r.t. the secondary GPT header not being at the end of the disk since this is something that would happen any time a GPT partitioned image is written to media of different size than the image itself.

When it comes to the dracut/initramfs issue, if it expects the medium to be formatted/partitioned in a certain way and this given ISOHYBRID format is incompatible with expectations, that is outside Limine's (or xorriso's) responsibility at that point.

**Comment by mobinmob** [comment]
> The reference xorriso command (as provided in Limine's own README) produces an ISOHYBRID using GPT + a protective MBR (as well as the actual ISO filesystem). This command is actually very similar to (almost taken right from) the one used by GRUB2 in their `grub-mkrescue` utility.
> 
> Given that I have hardly ever heard reports of an ISO generated in such matter not booting on any given system, I am inclined to believe that the modifications to the reference xorriso command made by @mobinmob are (at least partly) to blame, and I'd suggest sticking as closely as possible to the reference xorriso "incantation" to avoid this sort of issues.
> 

I changed the command to be as close to the reference in the limine documentation, thanks :)
It does not seem to make any difference regarding to the issue, unfortunatelly.
On some conditions, dracut cannot boot from the partition labeled VOID_ISO. There are two of them recognised, with the same UUID - at least that is how they appear in the blkid output at the dracut shell.

**Comment by mintsuki** [comment]
@mobinmob Thanks for applying my suggestion. Mind you, it wasn't intended as a fix for the dracut issue, but rather as a fix for @0x5c 's claim that Limine failed to boot on their Pentium 4 system.

**Pull request closed without merging**